src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+  return default_bp_niters(undirected_graph(underlying_graph(g)))
+end
+function message_tensor_diff(mt_a::Vector{ITensor}, mt_b::Vector{ITensor})
+  LHS, RHS = contract(mt_a), contract(mt_b)
+  return 0.5 *
+         norm((denseblocks(LHS) / sum(diag(LHS))) - (denseblocks(RHS) / sum(diag(RHS))))
+end
+
+struct BeliefPropagationCache{PTN,MTS}
+  partitioned_itensornetwork::PTN
+  message_tensors::MTS
+end
+
+#Constructors...
+function BeliefPropagationCache(ptn::PartitionedGraph; mt_storage=default_mt_storage)
+  mts = default_mt_storage(ptn)
+  return BeliefPropagationCache(ptn, mts)
+end
+
+function BeliefPropagationCache(tn::ITensorNetwork, partitioned_vertices; kwargs...)
+  ptn = PartitionedGraph(tn, partitioned_vertices)
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+partitioned_itensornetwork(bpc::BeliefPropagationCache) = bpc.partitioned_itensornetwork
+message_tensors(bpc::BeliefPropagationCache) = bpc.message_tensors
+function tensornetwork(bpc::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bpc))
+end
+function NamedGraphs.partitioned_graph(bpc::BeliefPropagationCache)
+  return partitioned_graph(partitioned_itensornetwork(bpc))
+end
+
+function initial_message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  ptn = partitioned_itensornetwork(bpc)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  inds_e = commoninds(src_e_itn, dst_e_itn)
+  return mt_constructor(inds_e)
+end
+
+function message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  mts = message_tensors(bpc)
+  return haskey(mts, edge) ? mts[edge] : initial_message_tensor(bpc, edge; mt_constructor)
+end
+function message_tensors(
+  bpc::BeliefPropagationCache, edges::Vector{PartitionEdge}; kwargs...
+)
+  return [message_tensor(bpc, edge; kwargs...) for edge in edges]
+end
+
+function copy(bpc::BeliefPropagationCache)
+  return BeliefPropagationCache(
+    copy(partitioned_itensornetwork(bpc)), copy(message_tensors(bpc))
+  )
+end
+
+function default_bp_niters(bpc::BeliefPropagationCache)
+  return default_bp_niters(partitioned_graph(bpc))
+end
+function default_edge_sequence(bpc::BeliefPropagationCache)
+  return default_edge_sequence(partitioned_itensornetwork(bpc))
+end
+
+"""
+Compute message tensor as product of incoming mts and local state
+"""
+function updated_message_tensor(
+  bpc::BeliefPropagationCache,
+  edge::PartitionEdge;
+  contractor=default_contractor,
+  contractor_kwargs=default_contractor_kwargs(),
+  mt_constructor=default_mt_constructor,
+)
+  ptn = partitioned_itensornetwork(bpc)
+  pb_edges = partitionedges(ptn, boundary_edges(ptn, vertices(ptn, src(edge)); dir=:in))
+
+  incoming_messages = [
+    message_tensor(bpc, e_in; mt_constructor) for e_in in setdiff(pb_edges, [reverse(edge)])
+  ]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
```

Comment:
Probably best to factorize this out into a function `incoming_messages(bpc::BeliefPropagationCache,
  edge::PartitionEdge)`.

Probably also helpful to have a function `incoming_messages(bpc::BeliefPropagationCache, vertex)` which `incoming_messages(bpc::BeliefPropagationCache,
  edge::PartitionEdge)` could make use of.

Reply:
Thinking about this it's probably best to have `boundary_partition_edges(ptn::PartitionedGraph, vertices::Vector{<:PartitionVertex}; dir)` defined in `NamedGraphs`? Then those functions will be fairly simple in the BP code here?

Reply:
Only issue with `incoming_messages(bpc::BeliefPropagationCache, vertex)` is it wouldn't filter out the message tensor incoming from `dst` to `src` and we wouldn't want to have to make a comparison between `messages` to do that

Reply:
Good point about using `incoming_messages(bpc::BeliefPropagationCache, vertex)` directly in `incoming_messages(bpc::BeliefPropagationCache, edge::PartitionEdge)`.

A general `boundary_partition_edges(ptn::PartitionedGraph, vertices::Vector{<:PartitionVertex}; dir)` in `NamedGraphs.jl` makes sense.

Reply:
We could write something like `incoming_messages(bpc::BeliefPropagationCache, vertices::Vector{<:PartitionVertex}; ignore_edges = PartitionEdge[])` which ignores some subset of edges? This would help unify things I think?

Reply:
That sounds like a good approach.